{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import malariagen_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\nconst JS_MIME_TYPE = 'application/javascript';\n  const HTML_MIME_TYPE = 'text/html';\n  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  const CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    const script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    const cell = handle.cell;\n\n    const id = cell.output_area._bokeh_element_id;\n    const server_id = cell.output_area._bokeh_server_id;\n    // Clean up Bokeh references\n    if (id != null && id in Bokeh.index) {\n      Bokeh.index[id].model.document.clear();\n      delete Bokeh.index[id];\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd_clean, {\n        iopub: {\n          output: function(msg) {\n            const id = msg.content.text.trim();\n            if (id in Bokeh.index) {\n              Bokeh.index[id].model.document.clear();\n              delete Bokeh.index[id];\n            }\n          }\n        }\n      });\n      // Destroy server and session\n      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd_destroy);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    const output_area = handle.output_area;\n    const output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      const bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      const script_attrs = bk_div.children[0].attributes;\n      for (let i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      const toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    const events = require('base/js/events');\n    const OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.1.1.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ag3 = malariagen_data.Ag3(pre=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subsetting to segregating sites: ⠏ (0:00:01.62) count alleles\n",
      "Subsetting to segregating sites: ⠴ (0:00:02.14) ascertain segregating  sites\n",
      "Subsetting to segregating sites: ⠦ (0:00:02.22) ascertained 1,089,159 sites\n",
      "thin sites                                      \n",
      "thinned to 2,002 sites\n",
      "Set up dataset\n",
      "[########################################] | 100% Completed | 103.94 ms\n",
      "Ensure genotypes vary                         \n",
      "final no. variants 138\n",
      "Load final data\n",
      "write plink files to /Users/dennistpw/Projects/malariagen-data-python//2L:100000-2000000.2000.0.0.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/Users/dennistpw/Projects/malariagen-data-python//2L:100000-2000000.2000.0.0.0'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ag3.biallelic_snps_to_plink(results_dir='/Users/dennistpw/Projects/malariagen-data-python/',\n",
    "                            region='2L:100000-2000000',\n",
    "                            n_snps=2000,\n",
    "                            sample_sets='AG1000G-AO',\n",
    "                            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "from typing import Optional, Tuple\n",
    "\n",
    "import allel  # type: ignore\n",
    "import numpy as np\n",
    "import os\n",
    "#import bed_reader\n",
    "from dask.diagnostics import ProgressBar\n",
    "# import bokeh.plotting\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import allel\n",
    "region='2L:100000-2000000'\n",
    "n_snps=2000\n",
    "sample_sets='AG1000G-AO'\n",
    "ds_snps = ag3.snp_calls(\n",
    "            sample_sets='AG1000G-AO',\n",
    "            region=region)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.DataArray &#x27;variant_allele&#x27; (variants: 1677564, alleles: 4)&gt;\n",
       "array([[b&#x27;C&#x27;, b&#x27;A&#x27;, b&#x27;T&#x27;, b&#x27;G&#x27;],\n",
       "       [b&#x27;G&#x27;, b&#x27;A&#x27;, b&#x27;C&#x27;, b&#x27;T&#x27;],\n",
       "       [b&#x27;C&#x27;, b&#x27;A&#x27;, b&#x27;T&#x27;, b&#x27;G&#x27;],\n",
       "       ...,\n",
       "       [b&#x27;A&#x27;, b&#x27;C&#x27;, b&#x27;T&#x27;, b&#x27;G&#x27;],\n",
       "       [b&#x27;A&#x27;, b&#x27;C&#x27;, b&#x27;T&#x27;, b&#x27;G&#x27;],\n",
       "       [b&#x27;A&#x27;, b&#x27;C&#x27;, b&#x27;T&#x27;, b&#x27;G&#x27;]], dtype=&#x27;|S1&#x27;)\n",
       "Coordinates:\n",
       "    variant_position  (variants) int32 100000 100001 100002 ... 1999999 2000000\n",
       "    variant_contig    (variants) uint8 1 1 1 1 1 1 1 1 1 1 ... 1 1 1 1 1 1 1 1 1\n",
       "Dimensions without coordinates: variants, alleles</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.DataArray</div><div class='xr-array-name'>'variant_allele'</div><ul class='xr-dim-list'><li><span>variants</span>: 1677564</li><li><span>alleles</span>: 4</li></ul></div><ul class='xr-sections'><li class='xr-section-item'><div class='xr-array-wrap'><input id='section-2e9741d7-318f-4379-9ae6-c9b532dff3c1' class='xr-array-in' type='checkbox' checked><label for='section-2e9741d7-318f-4379-9ae6-c9b532dff3c1' title='Show/hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-array-preview xr-preview'><span>b&#x27;C&#x27; b&#x27;A&#x27; b&#x27;T&#x27; b&#x27;G&#x27; b&#x27;G&#x27; b&#x27;A&#x27; b&#x27;C&#x27; ... b&#x27;T&#x27; b&#x27;G&#x27; b&#x27;A&#x27; b&#x27;C&#x27; b&#x27;T&#x27; b&#x27;G&#x27;</span></div><div class='xr-array-data'><pre>array([[b&#x27;C&#x27;, b&#x27;A&#x27;, b&#x27;T&#x27;, b&#x27;G&#x27;],\n",
       "       [b&#x27;G&#x27;, b&#x27;A&#x27;, b&#x27;C&#x27;, b&#x27;T&#x27;],\n",
       "       [b&#x27;C&#x27;, b&#x27;A&#x27;, b&#x27;T&#x27;, b&#x27;G&#x27;],\n",
       "       ...,\n",
       "       [b&#x27;A&#x27;, b&#x27;C&#x27;, b&#x27;T&#x27;, b&#x27;G&#x27;],\n",
       "       [b&#x27;A&#x27;, b&#x27;C&#x27;, b&#x27;T&#x27;, b&#x27;G&#x27;],\n",
       "       [b&#x27;A&#x27;, b&#x27;C&#x27;, b&#x27;T&#x27;, b&#x27;G&#x27;]], dtype=&#x27;|S1&#x27;)</pre></div></div></li><li class='xr-section-item'><input id='section-dbfcad9a-4510-4fd9-91ec-4d4929d5d9b4' class='xr-section-summary-in' type='checkbox'  checked><label for='section-dbfcad9a-4510-4fd9-91ec-4d4929d5d9b4' class='xr-section-summary' >Coordinates: <span>(2)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>variant_position</span></div><div class='xr-var-dims'>(variants)</div><div class='xr-var-dtype'>int32</div><div class='xr-var-preview xr-preview'>100000 100001 ... 1999999 2000000</div><input id='attrs-06e0eefd-0ddd-41bb-8b22-26b55744d7ad' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-06e0eefd-0ddd-41bb-8b22-26b55744d7ad' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-bc35d42b-a07c-4520-83ce-d73d2f42f60e' class='xr-var-data-in' type='checkbox'><label for='data-bc35d42b-a07c-4520-83ce-d73d2f42f60e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 100000,  100001,  100002, ..., 1999998, 1999999, 2000000],\n",
       "      dtype=int32)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>variant_contig</span></div><div class='xr-var-dims'>(variants)</div><div class='xr-var-dtype'>uint8</div><div class='xr-var-preview xr-preview'>1 1 1 1 1 1 1 1 ... 1 1 1 1 1 1 1 1</div><input id='attrs-3d60d38a-b96d-45d9-8350-3854430d2e85' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3d60d38a-b96d-45d9-8350-3854430d2e85' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1f28c15c-70c4-42f9-86ab-d1795500b520' class='xr-var-data-in' type='checkbox'><label for='data-1f28c15c-70c4-42f9-86ab-d1795500b520' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([1, 1, 1, ..., 1, 1, 1], dtype=uint8)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-6026316a-9257-4c3e-990e-9f52bc5b98a8' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-6026316a-9257-4c3e-990e-9f52bc5b98a8' class='xr-section-summary'  title='Expand/collapse section'>Indexes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'></ul></div></li><li class='xr-section-item'><input id='section-668b96dd-1ff1-4c63-9323-47b63a00f0a3' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-668b96dd-1ff1-4c63-9323-47b63a00f0a3' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.DataArray 'variant_allele' (variants: 1677564, alleles: 4)>\n",
       "array([[b'C', b'A', b'T', b'G'],\n",
       "       [b'G', b'A', b'C', b'T'],\n",
       "       [b'C', b'A', b'T', b'G'],\n",
       "       ...,\n",
       "       [b'A', b'C', b'T', b'G'],\n",
       "       [b'A', b'C', b'T', b'G'],\n",
       "       [b'A', b'C', b'T', b'G']], dtype='|S1')\n",
       "Coordinates:\n",
       "    variant_position  (variants) int32 100000 100001 100002 ... 1999999 2000000\n",
       "    variant_contig    (variants) uint8 1 1 1 1 1 1 1 1 1 1 ... 1 1 1 1 1 1 1 1 1\n",
       "Dimensions without coordinates: variants, alleles"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds_snps['variant_allele'].compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count alleles\n",
      "ascertain segregating  sites\n",
      "ascertained 1,089,159 sites\n"
     ]
    }
   ],
   "source": [
    "gt = ds_snps[\"call_genotype\"].data.compute()\n",
    "print(\"count alleles\")\n",
    "with ProgressBar():\n",
    "    ac = allel.GenotypeArray(gt).count_alleles(max_allele=3)\n",
    "    print(\"ascertain segregating  sites\")\n",
    "    n_chroms = ds_snps.dims[\"samples\"] * 2\n",
    "    an_called = ac.sum(axis=1)\n",
    "    an_missing = n_chroms - an_called\n",
    "    min_ref_ac = 0\n",
    "    max_ref_ac = n_chroms - 0\n",
    "    # here we choose segregating sites\n",
    "    loc_sites = (\n",
    "        (ac[:, 0] >= min_ref_ac)\n",
    "        & (ac[:, 0] <= max_ref_ac)\n",
    "        & (an_missing <= 0)\n",
    "    )\n",
    "print(f\"ascertained {np.count_nonzero(loc_sites):,} sites\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set up dataset\n",
      "[########################################] | 100% Completed | 698.99 ms\n",
      "Ensure genotypes vary\n",
      "final no. variants 67780\n",
      "Load final data\n"
     ]
    }
   ],
   "source": [
    "#set up dataset with required vars for plink conversion\n",
    "print(\"Set up dataset\")\n",
    "ds_snps_asc = (\n",
    "    ds_snps[[\"variant_contig\", \"variant_position\", \"variant_allele\", \"sample_id\", \"call_genotype\"]]\n",
    "    .isel(alleles=slice(0, 2))\n",
    "    .sel(variants=loc_sites)\n",
    "    )\n",
    "\n",
    "#compute gt ref counts\n",
    "gt_asc = ds_snps_asc[\"call_genotype\"].data.compute()\n",
    "gn_ref = allel.GenotypeDaskArray(gt_asc).to_n_ref(fill=-127) #what does the fill do\n",
    "with ProgressBar():\n",
    "    gn_ref = gn_ref.compute()\n",
    "\n",
    "print(\"Ensure genotypes vary\")\n",
    "loc_var = np.any(gn_ref != gn_ref[:, 0, np.newaxis], axis=1)\n",
    "print(f\"final no. variants {np.count_nonzero(loc_var)}\")\n",
    "\n",
    "\n",
    "print(\"Load final data\")\n",
    "with ProgressBar():\n",
    "    ds_snps_final = (\n",
    "        ds_snps_asc[[\"variant_contig\", \"variant_position\", \"variant_allele\", \"sample_id\"]]\n",
    "        .isel(variants=loc_var)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def apply_allele_mapping(x, mapping, max_allele):\n",
    "    \"\"\"Transform an array x, where the columns correspond to alleles,\n",
    "    according to an allele mapping.\n",
    "\n",
    "    Values from columns in x will be move to different columns in the\n",
    "    output according to the allele index mapping given.\n",
    "\n",
    "    \"\"\"\n",
    "    n_sites = x.shape[0]\n",
    "    n_alleles = x.shape[1]\n",
    "    assert mapping.shape[0] == n_sites\n",
    "    assert mapping.shape[1] == n_alleles\n",
    "\n",
    "    # Create output array.\n",
    "    out = np.empty(shape=(n_sites, max_allele + 1), dtype=x.dtype)\n",
    "\n",
    "    # Iterate over sites.\n",
    "    for i in range(n_sites):\n",
    "        # Iterate over alleles.\n",
    "        for allele_index in range(n_alleles):\n",
    "            # Find the new index for this allele.\n",
    "            new_allele_index = mapping[i, allele_index]\n",
    "\n",
    "            # Copy data to the new allele index.\n",
    "            if new_allele_index >= 0 and new_allele_index <= max_allele:\n",
    "                out[i, new_allele_index] = x[i, allele_index]\n",
    "\n",
    "    return out\n",
    "\n",
    "def trim_alleles(ac):\n",
    "    \"\"\"Remap allele indices to trim out unobserved alleles.\n",
    "\n",
    "    The idea here is that our SNP data includes alleles which\n",
    "    may not ever be observed, or may not be observed within a\n",
    "    particular subset of the data. For convenience, it is useful\n",
    "    to retain only alleles which are observed.\n",
    "\n",
    "    Here we use a given set of allele counts (ac) to identify\n",
    "    observed alleles and derive a mapping from old allele\n",
    "    indices to a new set of allele indices.\n",
    "\n",
    "    For example, if alleles 0 and 2 are observed at a given\n",
    "    site in ac, then 0 will be mapped to 0 and 2 will be mapped\n",
    "    to 1.\n",
    "\n",
    "    Similarly, if alleles 1 and 3 are observed at a given site\n",
    "    in ac, then 1 will be mapped to 0 and 3 will be mapped to 1.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    n_variants = ac.shape[0]\n",
    "    n_alleles = ac.shape[1]\n",
    "\n",
    "    # Create the output array - this is an allele mapping array,\n",
    "    # that specifies how to recode allele indices.\n",
    "    mapping = np.empty((n_variants, n_alleles), dtype=np.int32)\n",
    "\n",
    "    # The value of -1 indicates that there is no mapping for\n",
    "    # a given allele. We fill with -1 so we can then set\n",
    "    # values where we do observe alleles and can define a\n",
    "    # mapping.\n",
    "    mapping[:] = -1\n",
    "\n",
    "    # Iterate over variants.\n",
    "    for i in range(n_variants):\n",
    "        # This will be the new index that we are mapping this allele to, if the\n",
    "        # allele is observed.\n",
    "        new_allele_index = 0\n",
    "\n",
    "        # Iterate over columns (alleles) in the input array.\n",
    "        for allele_index in range(n_alleles):\n",
    "            # Access the count for the jth allele.\n",
    "            c = ac[i, allele_index]\n",
    "            if c > 0:\n",
    "                # We have found a non-zero allele count, remap the allele.\n",
    "                mapping[i, allele_index] = new_allele_index\n",
    "\n",
    "                # Increment to be ready for the next allele.\n",
    "                new_allele_index += 1\n",
    "\n",
    "    return mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "region='2L:1001000-2009000',\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds = ag3.snp_calls(\n",
    "    region=region,\n",
    "    sample_sets=sample_sets\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6ffc36a40b02468e99c7c63e6c186a10",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Compute SNP allele counts:   0%|          | 0/20 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Perform an allele count.\n",
    "ac = ag3.snp_allele_counts(\n",
    "    region=region,\n",
    "    sample_sets=sample_sets\n",
    ")\n",
    "\n",
    "# Locate biallelic SNPs.\n",
    "loc_bi = allel.AlleleCountsArray(ac).is_biallelic()\n",
    "\n",
    "# Remap alleles to squeeze out unobserved alleles.\n",
    "ac_bi = ac[loc_bi]\n",
    "allele_mapping = trim_alleles(ac_bi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_bi = ds.isel(variants=loc_bi)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Any, Dict, List, Optional, Tuple, Union\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Start building a new dataset.\n",
    "coords: Dict[str, Any] = dict()\n",
    "data_vars: Dict[str, Any] = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store sample IDs.\n",
    "coords[\"sample_id\"] = (\"samples\",), ds_bi[\"sample_id\"].data\n",
    "\n",
    "# Store contig.\n",
    "coords[\"variant_contig\"] = (\"variants\",), ds_bi[\"variant_contig\"].data\n",
    "\n",
    "# Store position.\n",
    "coords[\"variant_position\"] = (\"variants\",), ds_bi[\"variant_position\"].data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ds_bi' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mdask\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01marray\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mda\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m variant_allele \u001b[38;5;241m=\u001b[39m \u001b[43mds_bi\u001b[49m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvariant_allele\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m.\u001b[39mdata\n\u001b[1;32m      3\u001b[0m variant_allele \u001b[38;5;241m=\u001b[39m variant_allele\u001b[38;5;241m.\u001b[39mrechunk((variant_allele\u001b[38;5;241m.\u001b[39mchunks[\u001b[38;5;241m0\u001b[39m], \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m))\n\u001b[1;32m      4\u001b[0m variant_allele_out \u001b[38;5;241m=\u001b[39m da\u001b[38;5;241m.\u001b[39mmap_blocks(\n\u001b[1;32m      5\u001b[0m     \u001b[38;5;28;01mlambda\u001b[39;00m block: apply_allele_mapping(block, allele_mapping, max_allele\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m),\n\u001b[1;32m      6\u001b[0m     variant_allele,\n\u001b[1;32m      7\u001b[0m     dtype\u001b[38;5;241m=\u001b[39mvariant_allele\u001b[38;5;241m.\u001b[39mdtype,\n\u001b[1;32m      8\u001b[0m     chunks\u001b[38;5;241m=\u001b[39m(variant_allele\u001b[38;5;241m.\u001b[39mchunks[\u001b[38;5;241m0\u001b[39m], [\u001b[38;5;241m2\u001b[39m]),\n\u001b[1;32m      9\u001b[0m )\n",
      "\u001b[0;31mNameError\u001b[0m: name 'ds_bi' is not defined"
     ]
    }
   ],
   "source": [
    "import dask.array as da\n",
    "variant_allele = ds_bi[\"variant_allele\"].data\n",
    "variant_allele = variant_allele.rechunk((variant_allele.chunks[0], -1))\n",
    "variant_allele_out = da.map_blocks(\n",
    "    lambda block: apply_allele_mapping(block, allele_mapping, max_allele=1),\n",
    "    variant_allele,\n",
    "    dtype=variant_allele.dtype,\n",
    "    chunks=(variant_allele.chunks[0], [2]),\n",
    ")\n",
    "data_vars[\"variant_allele\"] = (\"variants\", \"alleles\"), variant_allele_out\n",
    "\n",
    "# Store allele counts, transformed, so we don't have to recompute.\n",
    "#ac_out = apply_allele_mapping(ac_bi, allele_mapping, max_allele=1)\n",
    "#data_vars[\"variant_allele_count\"] = (\"variants\", \"alleles\"), ac_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_allele = ds_bi[\"variant_allele\"].data\n",
    "variant_allele = variant_allele.rechunk((variant_allele.chunks[0], -1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[b'T', b'A', b'C', b'G'],\n",
       "       [b'G', b'A', b'C', b'T'],\n",
       "       [b'T', b'A', b'C', b'G'],\n",
       "       ...,\n",
       "       [b'G', b'A', b'C', b'T'],\n",
       "       [b'C', b'A', b'T', b'G'],\n",
       "       [b'A', b'C', b'T', b'G']], dtype='|S1')"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_allele.compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[b'T', b'A', b'C', b'G'],\n",
       "       [b'G', b'A', b'C', b'T'],\n",
       "       [b'T', b'A', b'C', b'G'],\n",
       "       ...,\n",
       "       [b'G', b'A', b'C', b'T'],\n",
       "       [b'C', b'A', b'T', b'G'],\n",
       "       [b'A', b'C', b'T', b'G']], dtype='|S1')"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_allele = variant_allele.rechunk((variant_allele.chunks[0], -1))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "   ds = self.snp_calls(\n",
    "            region=region,\n",
    "            sample_sets=sample_sets,\n",
    "            sample_query=sample_query,\n",
    "            sample_indices=sample_indices,\n",
    "            site_mask=site_mask,\n",
    "            site_class=site_class,\n",
    "            cohort_size=cohort_size,\n",
    "            min_cohort_size=min_cohort_size,\n",
    "            max_cohort_size=max_cohort_size,\n",
    "            random_seed=random_seed,\n",
    "            inline_array=inline_array,\n",
    "            chunks=chunks,\n",
    "        )\n",
    "\n",
    "        with self._spinner(\"Prepare biallelic SNP calls\"):\n",
    "            # Subset to biallelic sites.\n",
    "            ds_bi = ds.isel(variants=loc_bi)\n",
    "\n",
    "            # Start building a new dataset.\n",
    "            coords: Dict[str, Any] = dict()\n",
    "            data_vars: Dict[str, Any] = dict()\n",
    "\n",
    "            # Store sample IDs.\n",
    "            coords[\"sample_id\"] = (\"samples\",), ds_bi[\"sample_id\"].data\n",
    "\n",
    "            # Store contig.\n",
    "            coords[\"variant_contig\"] = (\"variants\",), ds_bi[\"variant_contig\"].data\n",
    "\n",
    "            # Store position.\n",
    "            coords[\"variant_position\"] = (\"variants\",), ds_bi[\"variant_position\"].data\n",
    "\n",
    "            # Store alleles, transformed.\n",
    "            variant_allele = ds_bi[\"variant_allele\"].data\n",
    "            variant_allele = variant_allele.rechunk((variant_allele.chunks[0], -1))\n",
    "            variant_allele_out = da.map_blocks(\n",
    "                lambda block: apply_allele_mapping(block, allele_mapping, max_allele=1),\n",
    "                variant_allele,\n",
    "                dtype=variant_allele.dtype,\n",
    "                chunks=(variant_allele.chunks[0], [2]),\n",
    "            )\n",
    "            data_vars[\"variant_allele\"] = (\"variants\", \"alleles\"), variant_allele_out\n",
    "\n",
    "            # Store allele counts, transformed, so we don't have to recompute.\n",
    "            ac_out = apply_allele_mapping(ac_bi, allele_mapping, max_allele=1)\n",
    "            data_vars[\"variant_allele_count\"] = (\"variants\", \"alleles\"), ac_out\n",
    "\n",
    "            # Store genotype calls, transformed.\n",
    "            gt = ds_bi[\"call_genotype\"].data\n",
    "            gt_out = allel.GenotypeDaskArray(gt).map_alleles(allele_mapping)\n",
    "            data_vars[\"call_genotype\"] = (\n",
    "                (\n",
    "                    \"variants\",\n",
    "                    \"samples\",\n",
    "                    \"ploidy\",\n",
    "                ),\n",
    "                gt_out.values,\n",
    "            )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "malariagen_plink",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
